File: app/src/main/java/com/wyattconrad/cs_360weighttracker/model/Goal.java
package com.wyattconrad.cs_360weighttracker.model;

import androidx.room.ColumnInfo;
import androidx.room.Entity;
import androidx.room.ForeignKey;
import androidx.room.Index;
import androidx.room.PrimaryKey;

// Set the username as unique
// Set a foreign key to the user table
// Set the child column on delete method
@Entity(indices = {@Index(value = "user_id", unique = true)},
        foreignKeys = @ForeignKey(entity = User.class, parentColumns = "id",
        childColumns = "user_id", onDelete = ForeignKey.CASCADE))
public class Goal {

    // Create the fields for the database
    // Primary key and ID for the goal
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    private long mId;

    // Goal value
    @ColumnInfo(name = "goal")
    private double mGoal;

    // Date and time set
    @ColumnInfo(name = "date_time_set")
    private long mDateTimeSet;

    // User ID that set the goal, foreign key from user table
    @ColumnInfo(name = "user_id")
    private long mUserId;


    /**
     * Default constructor
     * @param goal
     * @param mUserId
     */
    public Goal(double goal, long mUserId) {
        mGoal = goal;
        mDateTimeSet = System.currentTimeMillis();
        this.mUserId = mUserId;
    }

    // Getters and Setters
    public long getId() {
        return mId;
    }

    // Setter for Id
    public void setId(long id) {
        mId = id;
    }

    // Getter for Goal
    public double getGoal() {
        return mGoal;
    }

    // Setter for Goal
    public void setGoal(double goal) {
        mGoal = goal;
    }

    // Getter for DateTimeSet
    public long getDateTimeSet() {
        return mDateTimeSet;
    }

    // Setter for DateTimeSet
    public void setDateTimeSet(long dateTimeSet) {
        mDateTimeSet = dateTimeSet;
    }

    // Getter for UserId
    public long getUserId() {
        return mUserId;
    }

    // Setter for UserId
    public void setUserId(long userid) {
        mUserId = userid;
    }

}

File: app/src/main/java/com/wyattconrad/cs_360weighttracker/model/Goal.kt
/*
 * Copyright (C) 2025 Wyatt Conrad. All rights reserved.
 *
 * This file is part of the CS-360 Weight Tracker project.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.wyattconrad.cs_360weighttracker.model

import androidx.room.ColumnInfo
import androidx.room.Entity
import androidx.room.ForeignKey
import androidx.room.Index
import androidx.room.PrimaryKey

/**
 * The goal entity
 * @author Wyatt Conrad
 * @version 1.0
 */
// Set the username as unique
// Set a foreign key to the user table
// Set the child column on delete method
@Entity(
    tableName = "goals",
    indices = [Index(value = arrayOf("user_id"), unique = true)],
    foreignKeys = [ForeignKey(entity = User::class, parentColumns = ["id"], childColumns = ["user_id"], onDelete = ForeignKey.CASCADE)]
)
data class Goal(

    // Create the id as the primary key
    @PrimaryKey(autoGenerate = true)
    @ColumnInfo(name = "id")
    var id: Long = 0,

    // Create the date time set
    @ColumnInfo(name = "date_time_set")
    val dateTimeSet: Long,

    // Create the goal
    @field:ColumnInfo(name = "goal")
    val goal: Double,

    // Create the user id
    @field:ColumnInfo(name = "user_id")
    val userId: Long

)
